### PR TITLE
Colab and Jupyter notebook support, Official PyPi server hosting, web support

### LINKED ISSUE #618: [Tracking] Python 3.9 support
Python 3.9 is not yet supported because many dependencies do not support it yet.

This is a list of all dependencies and whether they are Python 3.9-ready. (Unchecked entries with no details have not been checked yet)

## Core dependencies
- [x] alabaster
  - pure Python
- [x] argcomplete
  - pure Python
- [x] Babel
  - pure Python
  - appears to have addressed 3.9 compatibility issues already
- [x] certifi
  - explicitly declares Python 3.9 support
- [x] chardet
  - pure Python
  - no issues reported
- [x] click
  - pure Python
  - tests passed with Python 3.9 recently
- [x] cycler
  - pure Python
- [x] decorator
  - fixed Python 3.9 (and some Python 3.10) issues in January!
- [x] dill
  - pure Python
- [x] dnspython
  - pure Python
- [x] docutils
  - pure Python
- [ ] eventlet
  - possible issues: https://github.com/eventlet/eventlet/issues?q=3.9
- [ ] fiftyone
  - pure Python
  - setup.py needs to be updated
- [ ] fiftyone-brain
  - depends on other dependencies, e.g. scipy
  - pyarmor works with 3.9 now
- [x] fiftyone-db
- [x] fiftyone-gui
- [x] Flask
  - tests pass with 3.9
- [x] Flask-Cors
  - pure Python
- [x] Flask-SocketIO
  - pure Python
- [ ] future
  - pure Python
  - some test issues may need investigating: https://github.com/PythonCharmers/python-future/issues?q=3.9
- [x] glob2
  - pure Python
  - tested manually
- [ ] greenlet
  - C extensions
  - no Windows wheel for 3.9 yet (as of 0.4.17): https://pypi.org/project/greenlet/#files
- [x] idna
  - pure Python
  - tests pass on 3.9
- [x] imageio
  - pure Python
  - appears to have resolved upstream 3.9 issues
- [x] imagesize
  - pure Python
  - tested manually
- [x] itsdangerous
  - pure Python
  - tests pass
- [ ] Jinja2
  - pure Python
  - need to check tests
- [x] joblib
  - pure Python
  - appears to have resolved 3.9 issues
- [ ] kiwisolver
  - no 3.9 wheels yet
  - may be published soon: https://github.com/nucleic/kiwi/issues?q=is%3Aissue+3.9
- [ ] MarkupSafe
- [ ] matplotlib
- [ ] mongoengine
- [ ] monotonic
- [ ] networkx
- [x] numpy
  - 1.19.3 has 3.9 wheels!
- [ ] opencv-python-headless
- [ ] packaging
- [ ] Pillow
- [x] pip
- [ ] pkg-resources
- [ ] pprintpp
- [ ] psutil
- [ ] Pygments
- [ ] pymongo
- [ ] pyparsing
- [ ] python-dateutil
- [ ] python-engineio
- [ ] python-socketio
- [ ] pytz
- [ ] PyWavelets
- [ ] requests
- [ ] retrying
- [ ] scikit-image
- [ ] scikit-learn
- [ ] scipy
  - 1.5.3 has no 3.9 wheels
- [x] setuptools
- [ ] six
- [ ] snowballstemmer
- [ ] sortedcontainers
- [ ] Sphinx
- [ ] sphinxcontrib-applehelp
- [ ] sphinxcontrib-devhelp
- [ ] sphinxcontrib-htmlhelp
- [ ] sphinxcontrib-jsmath
- [ ] sphinxcontrib-qthelp
- [ ] sphinxcontrib-serializinghtml
- [ ] tabulate
- [ ] threadpoolctl
- [ ] tifffile
- [ ] tzlocal
- [ ] urllib3
- [ ] voxel51-eta
- [ ] websocket-client
  - apparently broken: https://github.com/websocket-client/websocket-client/issues?q=isalive
- [x] Werkzeug
  - tests pass
- [x] wheel
- [x] xmltodict
  - pure Python

## `pycocotools` dependencies

### LINKED ISSUE #705: [BUG] App UI errors after using label selection filter
After using the label search UI to select a label to view, some App features stop working.

For example, in the state below, the checkbox to toggle the label field on and off no longer works after selecting the `person` label via the UI:

Screen Shot 2020-12-04 at 1 17 57 PM

### PR DESCRIPTION
**Pull request by benjaminpkane**
## What changes are proposed in this pull request?

* Adds web support, e.g. opening the App directly from the browser when a remote session has been forwarded to your machine
* Adds support for Google Colab and Jupyter notebooks
* Changes builds to publish to the official pypi server
* Adds (limited) 3.9 support
  * tensorflow does not support 3.9
  * scikit-image and scikit-learn do not offer built 3.9 wheels. Release candidates for dot releases with 3.9 support are up, so they should be available soon

I am proposing the simple install no longer includes the app, but includes web support.
```
pip install fiftyone
```

And to install the App, as well
```
pip install fiftyone[desktop]
```

Fixes #705. Previously, a React hook would begin to repeatedly update state when the label filter was used and freeze up the window

Closes #618 

TODOs:
* [x] Video is not fully functional at the moment. It looks like Colab strips response headers in its proxy so range requests are not handled by HTML5 video elements, i.e. the entire video has to load and seeking behavior does not work. Looking into it
* [x] Finish github actions updates
* [x] Finish App packaging updates
* [x] Update various help outputs
* [x] Update documentation
* [x] Test, test, test

## How is this patch tested? If it is not, please explain why.

Locally

### Is this a user-facing change that should be mentioned in the release notes?

-   [ ] No. You can skip the rest of this section.
-   [x] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

See forthcoming release notes.

### What areas of FiftyOne does this PR affect?

-   [x] App: FiftyOne application changes
-   [x] Build: Build and test infrastructure changes
-   [ ] Core: Core `fiftyone` Python library changes
-   [x] Documentation: FiftyOne documentation changes
-   [ ] Other

### CODE REVIEW

**Comment by benjaminpkane** [comment]
Ready for review

File: fiftyone/core/config.py [hunk 1]
```diff
@@ -57,13 +57,22 @@ def __init__(self, d):
...
+            default="browser",
+        )
         self._show_progress_bars = None  # declare
         self.show_progress_bars = self.parse_bool(
             d,
             "show_progress_bars",
             env_var="FIFTYONE_SHOW_PROGRESS_BARS",
             default=True,
         )
+        self.do_not_track = self.parse_bool(
```
Comment (brimoor) [inline fiftyone/core/config.py:73]: Suggestion: add an FAQ question that references the existence of this parameter and describes how we track (anonymous) usage

File: fiftyone/core/session.py [hunk 2]
```diff
@@ -35,7 +43,7 @@
 _subscribed_sessions = defaultdict(set)
 
 
-def launch_app(dataset=None, view=None, port=5151, remote=False):
+def launch_app(dataset=None, view=None, port=5151, remote=False, window=None):
```
Comment (brimoor) [inline fiftyone/core/session.py:46]: Suggestion: always print a message when `launch_app()` runs that explains what to expect:

- `desktop=False` (browser): open localhost:<port> in your browser
- `desktop=True` (desktop): FiftyOne desktop App has launched
- notebook: run `session.show()` to open the App in a cell

File: fiftyone/utils/uid.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+_TRACKING_NOTE = """
+FiftyOne tracks UUID based import usage by default. We are a small team looking
+to grow, and usage data is critical to that end.
+
+To disable tracking, use the `FIFTYONE_DO_NOT_TRACK=true` environment
+variable.
+"""
+
+
+def _get_user_id():
```
Comment (brimoor) [inline fiftyone/utils/uid.py:22]: Leaving myself a note to investigate when this message is printed during normal usage

**Review by brimoor — CHANGES_REQUESTED** [review]
Awesome stuff! I jotted down the comments I had that we discussed online.

The one other comment is to update the main `README.md` with:

(a) Updated instructions for performing a source install such that one can launch the App via either browser (or desktop, if possible?)  and develop non-App code

(b) Updated instructions for developing the App

**Comment by benjaminpkane** [comment]
Thanks for the thorough feedback! All comments have been addressed.

**Comment by brimoor** [comment]
@benjaminpkane Now that people are very likely to see the "You are not connected to a session" page in the App (eg whenever they exist a Python shell with a browser session open), can you update the text there?

Here's my suggestion for the text:

**Local sessions**

~~The following demonstrates how to connect to a local session from a python shell~~

Here's how to connect to a local session from Python:

```py
import fiftyone as fo

# Load your FiftyOne dataset
dataset = fo.load_dataset(...)

# Launch the app
session = fo.launch_app(dataset)
```

Screen Shot 2020-12-17 at 11 37 51 AM

**Remote sessions**

You can work with data on a remote machine by launching a remote App session and connecting to it from your local machine. [See this page](https://voxel51.com/docs/fiftyone/user_guide/app.html#remote-sessions) for more information.

On your remote machine

```py
import fiftyone as fo

# Load your FiftyOne dataset
dataset = fo.load_dataset(...)

# Launch a remote App instance that you'll connect to from your local machine
session = fo.launch_app(dataset, remote=True, port=XXXX)
```

On your local machine

```shell
# Option 1
# Use the CLI to configure port forwarding
# Connect to first remote session
fiftyone app connect \
    --destination <username>@<remote-ip-address> \
    --port XXXX --local-port 5151

# Option 2
# Manually configure port forwarding
ssh -N -L 5151:127.0.0.1:XXXX <username>@<remote-ip-address>
```

~~Port configuration~~

~~The default FiftyOne port is 5151. You can configure at anytime using the settings tab.~~

Screen Shot 2020-12-17 at 11 38 06 AM

**Comment by brimoor** [comment]
^the `5151` in the above example would ideally dynamically match the localhost port in browser mode. I guess in desktop mode, the Port configuration message at the bottom would still make sense.

**Comment by benjaminpkane** [comment]
Ok, ran into an issue which added some time to this. Here is the new setup page. The instructions use the active port number.

Screenshot from 2020-12-17 10-57-26
Screenshot from 2020-12-17 10-57-11

**Comment by benjaminpkane** [comment]
Let me add a different notebook instructions page. If they restart the kernel, the setup page is shown.

**Comment by benjaminpkane** [comment]
Notebook setup:

**Comment by brimoor** [comment]
Is it easy to change the hyperlink styling of `this page` so it is clear that it is a link?

Not critical by any means, but is it easy to do some basic syntax highlighting on the code snippets?

Great idea on a different notebook startup screen. Since you'd only see this in a cell that already has a `Session` defined, what about changing the help text to simply say:

```
Notebook sessions
-------------------

Re-run this cell to launch the App.
```

**Comment by benjaminpkane** [comment]
Right, I'll fix the link.

On the second point, If a notebook app cannot connect (like when the kernel is restarted) is means that `session` object does not exist.

**Comment by brimoor** [comment]
> On the second point, If a notebook app cannot connect (like when the kernel is restarted) is means that the `session` object does not exist.

Yeah, what I meant is that there must have already been code written (but not yet run) in the cell in order for there to be an unconnected App window there. So all the user needs to do is rerun the cell (and prior ones, likely) and then they'll have both their `session` object and their App window

**Comment by benjaminpkane** [comment]
Ok, I changed the text to:

```
Re-run the cell that created the session shown here
```

**Comment by brimoor** [comment]
Couple small comments per live discussion:

- Make `session.show()` call `session.refresh()` internally so that any dataset changes will be reflected when the App reloads

- Tweak link style (color and/or underline?) on "click here" so that it is a more obvious click target:

Screen Shot 2020-12-17 at 3 26 12 PM

**Comment by brimoor** [comment]
Oh and also I noticed that `session.show(height=100)` didn't seem to have any effect (window was still full width).

This isn't that critical (always full width in notebooks seems best anyway), just observing.

File: fiftyone/core/session.py [hunk 4]
```diff
@@ -6,15 +6,24 @@
...
+import fiftyone as fo
+import fiftyone.constants as focn
 import fiftyone.core.dataset as fod
 import fiftyone.core.client as foc
+import fiftyone.core.context as focx
 import fiftyone.core.service as fos
 from fiftyone.core.state import StateDescription
 
 
+html_escape = html.escape
```
Comment (brimoor) [inline fiftyone/core/session.py:25]: unnecessary?

Comment (benjaminpkane) [reply fiftyone/core/session.py:25]: Yes, will remove

**Comment by benjaminpkane** [comment]
Comments addressed

**Review by brimoor — APPROVED** [review]
Amazing stuff! Everything I've tried is working smoothly 🚀 🔥
